Add example support to SchemaType

steffann opened the pull request:
In schema.dtd the example tag is allowed inside the schema tag. This updates the parser to support it.

mgedmin commented:
I don't see any tests?

steffann commented:
I added example support in some more useful places

steffann commented:
Good point about the tests, I'll write them asap

freddrake commented:
- Documentation is missing (I keep meaning to convert to Sphinx, but haven't gotten to it; sorry!).

- What should an ``<example>`` contain for a section or multisection?  Should it be the content of a section, or one or more entire sections, including the enclosing ``<abc>`` ... ``</abc>`` markup?

- Examples should probably be validated (at least optionally).

steffann commented:
I didn't add the `<example>`, I only updated the code to match the dtd and vice-versa. I now notice that there is a difference between the dtd and the documentation. I don't see any use for `<example>` for section and multisection, and the documentation also doesn't document it. I'll update the dtd to match the documentation. I'll update the documentation to include `<example>` for sectiontypes, because that makes as much sense as supporting it for the schema.

steffann commented:
I would leave the `<example>` as free text for now as people might already be using it and the documentation explicitly mentions it will only be used for documentation purposes.

freddrake commented:
Yeah, I guess I'd forgotten about having added ``<example>`` way back.

I've been using a bit of RAML lately; that has rather decent example support.

steffann commented:
Thanks for that pointer, RAML looks useful.

About `<example>`: the current implementation is good enough for me. It can always be better (e.g. with optional validation, that is a good idea) but for now I would prefer to leave it as it is.

mgedmin commented:
I'd love to see some examples of the `<example>` tag.

mgedmin commented:
LGTM

freddrake commented:
@mgedmin: the documentation is in the **doc/** directory, but it's in the old-school LaTeX format still.

mgedmin commented:
I did a git grep for `<example>` and didn't find any examples of example tags, including in the doc/ subdirectory ... *shrug*

Never mind, it's not a prerequisite of every PR to fix *everything*.

freddrake commented:
Given that this is a straightforward implementation improvement rather than something completely new, I'm accepting this with the tests and no docs; I should have written docs back when I added the tag initially.

steffann commented:
Hi, sorry to open this ticket again, but I just realised that I removed too much in my previous commits. Originally ZConfig supported `<example>` in section and multisection even though the documentation didn't mention that. I thought it served no purpose so I removed them. One potential issue is of course that schema files that would parse correctly before now wouldn't. The other issue is that I actually found a use case for an example in a section so I would like them back :)

Patch included of course!
